### PR TITLE
Revamp the ESLint setup, closes LEA-1866

### PR DESCRIPTION
**Pull request by tigranpetrossian**
**Lots of words. Please read thoroughly and review at your own pace. Code changes are intentionally in separate commits, with relevant rule mentioned in the commit message body.** 

This PR revamps our ESLint setup:

* Upgrade ESLint to v9 
* Update the `eslint-config` package to extract out monorepo specific configurations and make it re-usable outside the monorepo.
* Add optional, configurable pre-commit and pre-push hooks so developers can run scripts of their choice automatically.

## eslint-config

The `eslint-config` package now holds less opinions about monorepo and/or Extension-specific setup. It’s rather an extension of few recommended presets we already used, with few important rule additions.
The package ships with two configs:

1. Base - `@leather.io/eslint-config`, this is the generic TS/JS setup suitable for all packages/projects
2. React - `@leather.io/eslint-config/react`, an additional preset used for packages/projects that use React. 

## Monorepo setup

ESLint v9 introduced a new [flat configuration pattern](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file), and a different mental model for monorepo setups, where the setup for the entire monorepo lives in a single file, instead of each package adding its own config file and extending the base config.

#### References:
* [The new config file](https://eslint.org/blog/2022/08/new-config-system-part-2/#the-new-config-file%3A-eslint.config.js)
* [Discussion on GitHub](https://github.com/eslint/eslint/discussions/16960#discussioncomment-5212286)

The [eslint-config.js](https://github.com/leather-io/mono/pull/765/files#diff-a32a0887ed9d1d707bbb3b845b7df7fd40e673c47e7b60a3ebd896b68d3b8839) for the monorepo now lives in the root and has the following structure:

```js
export default tseslint.config(
  {
    files: ['{packages,apps}/**/src/**/*.{ts,tsx}'],
    extends: [baseConfig],
  },
  {
    ignores: [...], // omitted for brevity
  },
  {
    name: 'ui',
    files: ['packages/ui/src/**/*.{ts,tsx}'],
    extends: [reactConfig],
  },
  {
    name: 'query',
    files: ['packages/query/src/**/*.{ts,tsx}'],
    extends: [reactConfig, pluginQuery.configs['flat/recommended']],
  },
  {
    name: 'mobile',
    files: ['apps/mobile/src/**/*.{ts,tsx}'],
    extends: [reactConfig, pluginLingui.configs['flat/recommended']],
    rules: {
      'lingui/no-unlocalized-strings': [
        'error',
        {
          ignoreFunction: [
            'Error',
            'console.log',
            'console.warn',
            'console.error',
            'it',
            'describe',
            'test',
          ],
        },
      ],
    },
  }
);

```

This way, the base config applies to all apps and packages by default. Packages that need additional setup have their own configuration objects specifying custom presets and rules.

### max-warnings 0

Rules with `warn` severity are no longer ignored by code checks and will need to be fixed before integrating. The rationale behind this:

1. The severity of certain rules is often arbitrary. Some `warn` rules could actually cause serious bugs and should not be warnings, e.g. react hook dependencies. There are, however too many of those coming from presets to override in our own setup one by one.
2. Many `warn` rules can be auto-fixed by `--fix`, e.g. missing `type` keyword, boolean casts, duplicate imports, etc.
3. The code triggering rules with `warn` severity usually never gets addressed, and logs for those linger forever in consoles and GitHub’s PR -> File tab.

### Running the lint script

the `lint` script now runs directly, instead of spawning multiple scripts using `turbo`. ESLint has its own caching mechanism that works well, and for day to day work developers will likely be running linting only on handful of files staged for commit (covered in Git hooks section).

## Notable preset/rules updates

* Removed extension-specific `globals` rules that were carried over during the setup. Extension currently doesn’t use the `eslint-config` package, the migration will be done separately.
* Removed `universal/native` preset. This preset included the *entire* [expo config](https://github.com/expo/eslint-config-universe), overriding lots of our rules and rules we rely on from recommended presets, some very important ones, like  missing hook dependencies. [According to authors](https://github.com/expo/expo/issues/28144#issuecomment-2049529131)  `universe` is an internal set of rules for them.  There’s also [eslint-config-expo](https://github.com/expo/expo/issues/28144#issuecomment-2049529131), which similarly includes lots of rules we have configured ourselves, and is therefore unnecessary.
* `@tanstack/eslint-plugin-query` was misconfigured and never worked (an easy mistake to make given the old config format). It will work from now on. Silenced the related code for the time being to avoid introducing bugs, as I’m not sure what behavior might depend on incorrect code to be incorrect. 
* Unused variables will now be caught by `eslint` instead of `tsc`. The ESLint rule for this is much more capable (destructured assignments, function arguments) and has uncovered lots of incorrect prop contracts.

The full list of notable rule additions **can be inspected by reviewing the PR commit by commit.** Where possible, the commit has a reference to the relevant rule in the body of the message.

### Rules from typescript-eslint typechecked config

There’s a separate list of typescript-specific rules form the [typechecked](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended-type-checked.ts)
config that I firmly believe should be re-enabled and enforced leveling up our software design and forcing us into more honest contracts. 

These are currently still disabled (as a separate list in the [config](https://github.com/leather-io/mono/pull/765/files#diff-a32a0887ed9d1d707bbb3b845b7df7fd40e673c47e7b60a3ebd896b68d3b8839)), fixing these will require a considerable amount of time, and is out of scope for this PR. I’ll follow up with this during one of upcoming engineering syncs.

## Cherry on top: optional git hooks
Added optional, configurable git hook setup where developers can choose which scripts they want to run with which git hooks. Those can be enabled by adding an `.env` file to the root of the repo:

```
# .env
PRE_COMMIT=format,lint
PRE_PUSH=test:unit,lint:filenames,typecheck,syncpack:lint
```

This is driven by a custom husky script that will block the git action and output the list of failed scripts.

Realistically, `PRE_COMMIT=format,lint` alone should be more than enough as the fastest and least annoying option. The `pre-commit` hook runs `eslint` and `prettier` through `lint-staged`, checking only staged files, and fixing them as necessary. This typically takes about 3-4 seconds.

## To be discussed:
* `deprecation` plugin was replaced by the new `'@typescript-eslint/no-deprecated` rule, but is currently disabled. Since all warnings now block integration, we need a better way of handling usages of deprecated dependencies, as it’s not always realistic to address those in a short amount of time. I’m open to ideas about how we can track and log existing deprecation warnings into actionable Linear issues.

### CODE REVIEW

File: packages/eslint-config/README.md [hunk 1]
```diff
@@ -12,65 +16,62 @@ This package contains `eslint` configuration to be shared across projects.
...
-      project: path.resolve(__dirname, './tsconfig.json'),
-    },
-
-    ...,
-  };
-  ```
+### Inside @leather.io/mono
+
+You don’t need to install or link this package within monorepo packages. The base config 
+will be applied to all packages and apps by default. To add custom configurations (e.g., React) for a new
```
Comment (kyranjamie) [inline packages/eslint-config/README.md:22]:
```suggestion
is applied to all packages and apps by default. To add custom configurations (e.g., React) for a new 
```

active voice

File: packages/ui/src/icons/logos/logo-mpc-portal.native.tsx [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
-export const LogoMpcPortal = forwardRef<Component, IconProps>(({ variant, ...props }, ref) => {
+export const LogoMpcPortal = forwardRef<Component, SvgProps>((props, ref) => {
   return (
     <Icon ref={ref} {...props}>
       <LogoMpcPortalLogo />
     </Icon>
   );
 });
+
+LogoMpcPortal.displayName = 'LogoMpcPortal';
```
Comment (kyranjamie) [inline packages/ui/src/icons/logos/logo-mpc-portal.native.tsx:15]: is there a plugin or build tool of some sort that does this automatically?

**Review by kyranjamie — APPROVED** [review]
Great work @tigranpetrossian this is fantastic. Appreciate you making the hooks configurable

File: packages/eslint-config/old.eslintrc.js [hunk 3]
```diff
@@ -55,35 +56,35 @@ module.exports = {
...
+    '@typescript-eslint/no-non-null-asserted-optional-chain': ['off'], // -
+    '@typescript-eslint/no-explicit-any': ['off'], // +
+    '@typescript-eslint/no-meaningless-void-operator': 'error', //+
+    '@typescript-eslint/ban-types': ['error'], // + (three new rules in the new version)
+    '@typescript-eslint/no-unnecessary-condition': 'off', // -
+    '@typescript-eslint/consistent-type-definitions': ['error', 'interface'], // +
+    '@typescript-eslint/array-type': ['error'], // +
+    'no-warning-comments': [0], // -
+
+    'react-hooks/rules-of-hooks': 'error', // +
```
[the remark this answers is not in the document]
Comment (tigranpetrossian) [reply packages/eslint-config/old.eslintrc.js]: Thanks, this was a leftover.

File: packages/ui/src/icons/logos/logo-mpc-portal.native.tsx (hunk 2 above)
Comment (tigranpetrossian) [reply packages/ui/src/icons/logos/logo-mpc-portal.native.tsx:15]: There isn't a `--fix` for this rule, but I vaguely remember seeing a babel plugin, I can look into that. Either way this isn't a particularly annoying rule, display names are only required for `forwardRef`s, and we only have a handful of those, mostly in the ui library. 
The repetitive icon definitions should be abstracted away with a HOC (I'm slowly working on this as part of icon system revamp).
cc @edgarkhanzadian

**Comment by pete-watters** [comment]
Great work @tigranpetrossian 👏

File: apps/mobile/src/app/(home)/secure-your-wallet.tsx [hunk 4]
```diff
@@ -24,7 +23,8 @@ export default function SecureYourWalletScreen() {
         style={{ paddingBottom: bottom + theme.spacing[5] }}
       >
         <AnimatedHeaderScreenLayout
-          rightTitleElement={<MoreInfoIcon onPress={() => {}} />}
+          // Hidden until linked
```
Comment (pete-watters) [inline apps/mobile/src/app/(home)/secure-your-wallet.tsx:26]: Maybe you can add the ticker reference here?

File: apps/mobile/src/features/send/send-form/components/send-form-amount-field.tsx [hunk 5]
```diff
@@ -6,6 +6,7 @@ import { z } from 'zod';
 import { SendFormBaseContext, useSendFormContext } from '../send-form-context';
 
 export function SendFormAmountField<T extends SendFormBaseContext<T>>() {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (pete-watters) [inline apps/mobile/src/features/send/send-form/components/send-form-amount-field.tsx:9]: Why do we need to disable lint for `formData`? 

I see `eslint` gives this error: `'formData' is assigned a value but only used as a type` . Maybe we can fix that instead?

File: apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx [hunk 6]
```diff
@@ -101,7 +101,7 @@ export function useSendFormBtc() {
 
       // Show an error toast here?
       if (!tx) throw new Error('Attempted to generate raw tx, but no tx exists');
-      // eslint-disable-next-line no-console, lingui/no-unlocalized-strings
+      // eslint-disable-next-line no-console
```
Comment (pete-watters) [inline apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx:104]: I think we should probably just get rid of the `console.log`s here

File: packages/ui/src/components/highlighting/clarity-prism.shared.ts [hunk 7]
```diff
@@ -104,11 +104,11 @@ buff|hash160|sha256|sha512|sha512/256|keccak256|true|false|none)' +
       lookbehind: true,
     },
     address: {
-      pattern: /([\s()])(?:\'[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41})(?=[()\s]|$)/,
+      pattern: /([\s()])(?:'[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41})(?=[()\s]|$)/,
```
Comment (pete-watters) [inline packages/ui/src/components/highlighting/clarity-prism.shared.ts:107]: Just double checking these are intentional changes?

File: apps/mobile/src/features/send/send-form/components/send-form-amount-field.tsx (hunk 5 above)
Comment (tigranpetrossian) [reply apps/mobile/src/features/send/send-form/components/send-form-amount-field.tsx:9]: I assume this is work in progress by @fbwoolf, definitely worth tightening it up, but probably out of scope for this PR.

File: apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx (hunk 6 above)
Comment (tigranpetrossian) [reply apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx:104]: Probably out of scope for this PR, I think this is still work in progress too.

**Comment by tigranpetrossian** [comment]
> Great work @tigranpetrossian 👏 . I checked out this branch and branched off it to test it out.
> 
> When I try and push it keeps giving me a `syncpack` error?
> 
> ```
> The following checks failed. See output above for more details.
> ────────────────────────────────────────────────────────────────────────
> ✖ syncpack
> ```
> 
> When I run `pnpm syncpack:fix` it says it's already valid 🤔
> 
> I trust it's working for you but thought I should mention it

Oh, my bad. Script names in .env should match that of package.json. So you should have `syncpack:lint` instead. I'll update the README.

File: apps/mobile/src/app/(home)/secure-your-wallet.tsx (hunk 4 above)
Comment (tigranpetrossian) [reply apps/mobile/src/app/(home)/secure-your-wallet.tsx:26]: Good call! I have a reminder hanging in the Slack thread for this for days now. Will make one.

File: apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx (hunk 6 above)
Comment (pete-watters) [reply apps/mobile/src/features/send/send-form/hooks/use-send-form-btc.tsx:104]: Just if you are updating the lint, may as well delete the log instead.

Now we've released a beta we should tighten up and get rid of any of this stuff

**Comment by pete-watters** [comment]
> > Great work @tigranpetrossian 👏 . I checked out this branch and branched off it to test it out.
> > When I try and push it keeps giving me a `syncpack` error?
> > ```
> > The following checks failed. See output above for more details.
> > ────────────────────────────────────────────────────────────────────────
> > ✖ syncpack
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When I run `pnpm syncpack:fix` it says it's already valid 🤔
> > I trust it's working for you but thought I should mention it
> 
> @pete-watters Oh, my bad. Script names in .env should match that of package.json. So you should have `syncpack:lint` instead. I'll update the README. Thanks for testing this.

No problem, thank you for the great work 🙇  I'm happy to see `husky` back as a first barrier to stop lint / CI errors

**Review by fbwoolf — APPROVED** [review]
Wow, nice updates @tigranpetrossian! 👍

**Comment by camerow** [comment]
>  Code changes are intentionally in separate commits, with relevant rule mentioned in the commit message body.

This made things particularly easy to follow, nice one @tigranpetrossian!

**Pull request merged**